lib/ipam/ipam_block_reader_writer.go

```diff
@@ -120,11 +128,15 @@ func (rw blockReaderWriter) findUnclaimedBlock(ctx context.Context, host string,
 		for subnet := blocks(); subnet != nil; subnet = blocks() {
 			// Check if a block already exists for this subnet.
 			log.Debugf("Getting block: %s", subnet.String())
-			if _, ok := exists[subnet.String()]; !ok {
+			if info, ok := exists[subnet.String()]; !ok {
 				log.Infof("Found free block: %+v", *subnet)
 				return subnet, nil
+			} else if info.affinity == host && info.numFree != 0 {
+				// Belongs to this host and has free allocations.
+				log.Debugf("Block %s already assigned to host, has free space", subnet.String())
+				return subnet, nil
 			}
-			log.Debugf("Block %s already exists", subnet.String())
+			log.Debugf("Block %s already exists, try another", subnet.String())
```

Comment:
really minor nit:  Might be nice to say in the log that assigned to different host or no free space.